File: case_management_backend/app_old/routers/cases.py
"""Case management API.

Read paths (`/api/cases`, `/summary`, `/facets`) back the Case Management
screen; the write paths back triage actions; `/ingest` is what the rule engine
calls when a control fails.

Route order matters: the literal paths (`/summary`, `/facets`, `/ingest`) are
declared before `/{case_id}`, otherwise FastAPI would match them as an id.
"""

from __future__ import annotations

import csv
import io
from datetime import datetime

from fastapi import APIRouter, Body, Depends, File, Form, Query, Response, UploadFile
from fastapi.responses import FileResponse
from sqlalchemy.orm import Session

from app import attachments, case_service, schemas
from app.case_service import CaseFilters
from app.db import get_db

router = APIRouter(prefix="/api/cases", tags=["cases"])

# Repeated across the list/summary/export endpoints — declared once so the three
# always accept exactly the same filter surface.
def filter_params(
    q: str | None = Query(None, description="Free text over reference, title, rule, batch, owner"),
    assurance: list[str] | None = Query(None, description="Assurance code or name, repeatable"),
    group: list[str] | None = Query(None, description="Assurance group, e.g. Revenue Assurance"),
    module: list[str] | None = Query(None, description="Sub-module / entity scope"),
    category: list[str] | None = Query(None, description="Issue type (rule primitive category)"),
    status: list[str] | None = Query(None),
    severity: list[str] | None = Query(None),
    origin: list[str] | None = Query(None),
    action: list[str] | None = Query(None),
    owner: list[str] | None = Query(None),
    rule_id: list[str] | None = Query(None, alias="ruleId"),
    stream: list[str] | None = Query(None),
    assignment: str = Query("all", pattern="^(all|assigned|unassigned|mine)$"),
    me: str | None = Query(None, description="Current user, required by assignment=mine"),
    date_from: datetime | None = Query(None, alias="dateFrom"),
    date_to: datetime | None = Query(None, alias="dateTo"),
    date_field: str = Query("createdAt", alias="dateField",
                            pattern="^(createdAt|detectedAt|updatedAt)$"),
    open_only: bool = Query(False, alias="openOnly"),
) -> CaseFilters:
    return CaseFilters(
        q=q,
        assurance=assurance or [],
        group=group or [],
        module=module or [],
        category=category or [],
        status=status or [],
        severity=severity or [],
        origin=origin or [],
        action=action or [],
        owner=owner or [],
        rule_id=rule_id or [],
        stream=stream or [],
        assignment=assignment,
        me=me,
        date_from=date_from,
        date_to=date_to,
        date_field=date_field,
        open_only=open_only,
    )


# --------------------------------------------------------------------------
# Reads
# --------------------------------------------------------------------------

@router.get("", response_model=schemas.CaseListResponse)
def list_cases(
    filters: CaseFilters = Depends(filter_params),
    page: int = Query(1, ge=1),
    page_size: int = Query(25, ge=1, le=500, alias="pageSize"),
    sort_by: str = Query("createdAt", alias="sortBy"),
    sort_dir: str = Query("desc", alias="sortDir", pattern="^(asc|desc)$"),
    db: Session = Depends(get_db),
) -> schemas.CaseListResponse:
    """Filtered, sorted, paginated case list."""
    return case_service.list_cases(db, filters, page, page_size, sort_by, sort_dir)


@router.get("/summary", response_model=schemas.SummaryResponse)
def summary(
    filters: CaseFilters = Depends(filter_params),
    db: Session = Depends(get_db),
) -> schemas.SummaryResponse:
    """Tile counts for the current filter scope."""
    return case_service.summarise(db, filters)


@router.get("/facets", response_model=schemas.FacetsResponse)
def facets(db: Session = Depends(get_db)) -> schemas.FacetsResponse:
    """Filter dropdown options with counts, derived from existing cases."""
    return case_service.facets(db)


@router.get("/export.csv")
def export_csv(
    filters: CaseFilters = Depends(filter_params),
    sort_by: str = Query("createdAt", alias="sortBy"),
    sort_dir: str = Query("desc", alias="sortDir", pattern="^(asc|desc)$"),
    db: Session = Depends(get_db),
) -> Response:
    """The filtered set as CSV — same filters as the list, no page cap."""
    result = case_service.list_cases(db, filters, page=1, page_size=500, sort_by=sort_by, sort_dir=sort_dir)
    columns = [
        ("reference", "Case"), ("created_at", "Created"), ("assurance_name", "Assurance"),
        ("module", "Module"), ("rule_id", "Rule"), ("rule_name", "Rule name"),
        ("rule_category", "Issue type"), ("title", "Title"), ("status", "Status"),
        ("severity", "Priority"), ("action", "Action"), ("owner", "Assigned to"),
        ("expected_value", "Expected"), ("actual_value", "Actual"), ("variance", "Variance"),
        ("affected_count", "Affected records"), ("estimated_impact", "Impact"),
        ("linked_batch", "Batch"), ("origin", "Origin"),
    ]
    buf = io.StringIO()
    writer = csv.writer(buf)
    writer.writerow([label for _, label in columns])
    for row in result.items:
        writer.writerow([
            # Timestamps go out in the same UTC ISO form as the JSON API, so a
            # spreadsheet and the UI never disagree about when a case was cut.
            "" if (v := getattr(row, key)) is None else (schemas.to_iso(v) if isinstance(v, datetime) else v)
            for key, _ in columns
        ])
    return Response(
        content=buf.getvalue(),
        media_type="text/csv",
        headers={"Content-Disposition": 'attachment; filename="cases.csv"'},
    )


# --------------------------------------------------------------------------
# Writes — rule engine
# --------------------------------------------------------------------------

@router.post("/ingest", response_model=schemas.CaseDetail, status_code=201)
def ingest(
    payload: schemas.RuleCaseIngest,
    response: Response,
    db: Session = Depends(get_db),
) -> schemas.CaseDetail:
    """Raise a case from a control run.

    Send `dedupeKey` (e.g. `UA001:2026-07-30:NODE-2`) to make retries safe — a
    repeat post returns the case that already exists instead of a duplicate.
    201 means a new case was cut, 200 that one already existed.
    """
    case, created = case_service.ingest_rule_case(db, payload)
    if not created:
        response.status_code = 200
    return case_service.get_detail(db, case.id)


@router.post("/ingest/bulk", response_model=schemas.BulkIngestResponse, status_code=201)
def ingest_bulk(
    payload: list[schemas.RuleCaseIngest] = Body(...),
    db: Session = Depends(get_db),
) -> schemas.BulkIngestResponse:
    """Raise many cases from one rule-engine cycle."""
    created: list[schemas.CaseRow] = []
    duplicates: list[schemas.CaseRow] = []
    for item in payload:
        case, was_created = case_service.ingest_rule_case(db, item)
        row = schemas.CaseRow.model_validate(case)
        row.mismatch_count = len(case.mismatches)
        (created if was_created else duplicates).append(row)
    return schemas.BulkIngestResponse(
        created=created, duplicates=duplicates,
        created_count=len(created), duplicate_count=len(duplicates),
    )


# --------------------------------------------------------------------------
# Writes — analyst
# --------------------------------------------------------------------------

@router.post("", response_model=schemas.CaseDetail, status_code=201)
def create_case(payload: schemas.CaseCreate, db: Session = Depends(get_db)) -> schemas.CaseDetail:
    """Raise a case manually (the Add Case dialog)."""
    case = case_service.create_case(
        db, payload, origin=payload.origin, reference=payload.reference,
        actor=payload.created_by or payload.owner or "analyst",
    )
    return case_service.get_detail(db, case.id)


@router.get("/{case_id}", response_model=schemas.CaseDetail)
def get_case(case_id: str, db: Session = Depends(get_db)) -> schemas.CaseDetail:
    """One case with its mismatch rows, notes and audit trail. Accepts the
    uuid or the CASE-#### reference."""
    return case_service.get_detail(db, case_id)


@router.patch("/{case_id}", response_model=schemas.CaseDetail)
def update_case(
    case_id: str, payload: schemas.CaseUpdate, db: Session = Depends(get_db)
) -> schemas.CaseDetail:
    """Partial update; every tracked change is written to the audit trail."""
    return case_service.update_case(db, case_id, payload)


@router.post("/{case_id}/assign", response_model=schemas.CaseDetail)
def assign_case(
    case_id: str, payload: schemas.AssignIn, db: Session = Depends(get_db)
) -> schemas.CaseDetail:
    """Assign (or, with an empty owner, unassign) a case."""
    return case_service.assign_case(db, case_id, payload)


@router.post("/{case_id}/status", response_model=schemas.CaseDetail)
def set_status(
    case_id: str, payload: schemas.StatusIn, db: Session = Depends(get_db)
) -> schemas.CaseDetail:
    """Move a case through its lifecycle, optionally with a closing note."""
    return case_service.set_status(db, case_id, payload)


@router.post("/{case_id}/comments", response_model=schemas.CommentOut, status_code=201)
def add_comment(
    case_id: str, payload: schemas.CommentIn, db: Session = Depends(get_db)
) -> schemas.CommentOut:
    """Add an investigation note."""
    return case_service.add_comment(db, case_id, payload)


@router.post("/{case_id}/insights", response_model=schemas.CaseDetail, status_code=201)
def add_insight(
    case_id: str, payload: schemas.InsightIn, db: Session = Depends(get_db)
) -> schemas.CaseDetail:
    """Pin an assistant reply to the case."""
    return case_service.add_insight(db, case_id, payload)


@router.get("/{case_id}/mismatches")
def list_mismatches(
    case_id: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(100, ge=1, le=1000, alias="pageSize"),
    status: str | None = Query(None),
    db: Session = Depends(get_db),
) -> dict:
    """The mismatch evidence behind a case, paginated."""
    return case_service.list_mismatches(db, case_id, page, page_size, status)


@router.post("/{case_id}/mismatches", response_model=schemas.CaseDetail, status_code=201)
def add_mismatches(
    case_id: str,
    payload: list[schemas.MismatchIn] = Body(...),
    db: Session = Depends(get_db),
) -> schemas.CaseDetail:
    """Attach further mismatch rows to an existing case."""
    return case_service.add_mismatches(db, case_id, payload)


@router.delete("/{case_id}", response_model=schemas.OkResponse)
def delete_case(case_id: str, db: Session = Depends(get_db)) -> schemas.OkResponse:
    """Delete a case and everything hanging off it, including its files."""
    case_service.delete_case(db, case_id)
    return schemas.OkResponse(message=f"Case {case_id} deleted")


# --------------------------------------------------------------------------
# Attachments (PDF evidence)
# --------------------------------------------------------------------------

@router.get("/{case_id}/attachments", response_model=list[schemas.AttachmentOut])
def list_attachments(case_id: str, db: Session = Depends(get_db)) -> list[schemas.AttachmentOut]:
    """Files attached to a case."""
    return case_service.list_attachments(db, case_id)


@router.post("/{case_id}/attachments", response_model=list[schemas.AttachmentOut], status_code=201)
async def upload_attachments(
    case_id: str,
    files: list[UploadFile] = File(..., description="PDF evidence"),
    uploaded_by: str = Form("", alias="uploadedBy"),
    db: Session = Depends(get_db),
) -> list[schemas.AttachmentOut]:
    """Attach PDF evidence to a case (multipart/form-data, field name `files`).

    Rejects anything that isn't a PDF — both the declared content type and the
    file's own magic bytes are checked — and anything over the size limit.
    """
    return case_service.add_attachments(db, case_id, files, uploaded_by)


@router.get("/{case_id}/attachments/{attachment_id}/download")
def download_attachment(
    case_id: str, attachment_id: str, db: Session = Depends(get_db)
) -> FileResponse:
    """Stream one attachment back with its original filename."""
    case = case_service.get_case_or_404(db, case_id)
    attachment = attachments.get_or_404(db, case.id, attachment_id)
    return FileResponse(
        attachments.resolve_path(attachment),
        media_type=attachment.content_type,
        filename=attachment.filename,
        headers={"X-Checksum-SHA256": attachment.checksum_sha256},
    )


@router.delete("/{case_id}/attachments/{attachment_id}", response_model=schemas.OkResponse)
def delete_attachment(
    case_id: str,
    attachment_id: str,
    actor: str = Query("system"),
    db: Session = Depends(get_db),
) -> schemas.OkResponse:
    """Remove an attachment and its stored file."""
    case_service.remove_attachment(db, case_id, attachment_id, actor)
    return schemas.OkResponse(message="Attachment deleted")

File: ra_demo/src/lib/cases.ts
// Case management client — talks to the FastAPI backend (app/routers/cases.py).
//
// Cases are raised two ways and both land in the same store: the rule engine
// posts to /api/cases/ingest when a control fails, and analysts create them
// from the Add Case dialog. Every case therefore carries the assurance, the
// sub-module, the rule that fired and its issue type — which is exactly what
// the filters below are built on.

// Same convention as the pipelines client: its own base so the cases service can
// live on a different host/port. Empty (as in .env.production) means same-origin,
// where nginx proxies /api.
//
// 127.0.0.1, not "localhost": on this machine localhost resolves to ::1 first,
// and the API binds IPv4, so a browser's first connection attempt is refused and
// the request shows up as a failed fetch with provisional headers. Naming the
// IPv4 address removes the ambiguity.
const API_BASE: string = import.meta.env.VITE_CASES_API_BASE ?? "http://127.0.0.1:8001";

// The signed-in demo analyst. Cases owned by this name populate "Self Assigned"
// and are what "Assign to me" writes.
export const CURRENT_ANALYST = "super_admin";

// ---------------------------------------------------------------------------
// Types — mirror the backend schemas (app/schemas.py), which serialise camelCase
// ---------------------------------------------------------------------------

export const STATUSES = ["Open", "In Progress", "Resolved", "Closed", "Cancelled"] as const;
export const SEVERITIES = ["low", "medium", "high", "critical"] as const;
export const ORIGINS = ["auto_detected", "analyst_raised"] as const;
export const ACTIONS = [
  "NA", "Escalated to carrier", "Adjusted & rebilled", "Waived",
  "Config fix raised", "Re-ingest requested", "Under review",
] as const;

export type CaseStatus = (typeof STATUSES)[number];
export type CaseSeverity = (typeof SEVERITIES)[number];

export interface CaseMismatch {
  id: string;
  position: number;
  recordRef: string;
  entity: string;
  subscriber: string;
  field: string;
  expectedValue: string | null;
  actualValue: string | null;
  delta: string | null;
  status: string;
  occurredAt: string | null;
  details: Record<string, unknown>;
}

export interface CaseComment {
  id: string;
  author: string;
  body: string;
  createdAt: string;
}

/** A file attached to a case as evidence. PDF only, enforced server-side. */
export interface CaseAttachment {
  id: string;
  caseId: string;
  filename: string;
  contentType: string;
  sizeBytes: number;
  checksumSha256: string;
  uploadedBy: string;
  createdAt: string;
  /** Server-built path; join with the API base to download. */
  downloadUrl: string;
}

export interface CaseActivity {
  id: string;
  actor: string;
  action: string;
  field: string | null;
  fromValue: string | null;
  toValue: string | null;
  note: string | null;
  createdAt: string;
}

export interface AssuranceCase {
  id: string;
  reference: string;
  title: string;
  description: string;

  // Which assurance raised it, and where inside that assurance.
  assuranceCode: string;
  assuranceName: string;
  assuranceGroup: string;
  module: string;
  subModule: string;

  // The control that fired. Null for analyst-raised cases.
  ruleId: string | null;
  ruleName: string | null;
  ruleCategory: string;   // issue type: Completeness, Reconciliation, …
  ruleRunId: string | null;

  origin: string;
  severity: string;
  status: string;
  action: string;
  owner: string;          // "" means unassigned

  stream: string;
  nodeId: string;
  sourceFeed: string;
  targetFeed: string;
  linkedBatch: string;
  linkedTxnId: string;

  // What the control measured — the headline mismatch.
  expectedValue: string | null;
  actualValue: string | null;
  variance: string | null;
  variancePct: number | null;
  threshold: string | null;
  estimatedImpact: number;
  affectedCount: number;

  evidence: { name: string; kind: string; size: string } | null;
  tags: string[];
  savedInsights: { id: string; body: string; at: string }[];
  details: Record<string, unknown>;

  detectedAt: string;
  createdAt: string;
  updatedAt: string;
  closedAt: string | null;

  mismatchCount: number;

  // Present on the detail response only.
  mismatches?: CaseMismatch[];
  comments?: CaseComment[];
  activities?: CaseActivity[];
  attachments?: CaseAttachment[];
}

export interface CaseListResponse {
  items: AssuranceCase[];
  total: number;
  page: number;
  pageSize: number;
  pageCount: number;
}

export interface CaseSummary {
  total: number;
  unassigned: number;
  assigned: number;
  byStatus: Record<string, number>;
  bySeverity: Record<string, number>;
  byAssurance: Record<string, number>;
  byCategory: Record<string, number>;
  openEstimatedImpact: number;
  affectedRecords: number;
}

export interface FacetValue {
  value: string;
  label: string;
  count: number;
}

export interface CaseFacets {
  assurances: FacetValue[];
  groups: FacetValue[];
  modules: FacetValue[];
  categories: FacetValue[];
  statuses: FacetValue[];
  severities: FacetValue[];
  origins: FacetValue[];
  actions: FacetValue[];
  owners: FacetValue[];
  rules: FacetValue[];
}

export interface CatalogAssurance {
  code: string;
  name: string;
  group: string;
  modules: string[];
}

export interface CatalogMeta {
  assurances: CatalogAssurance[];
  groups: string[];
  modules: string[];
  ruleCategories: string[];
  primaryRuleCategories?: string[];
  hiddenRuleCategories?: string[];
  categoryParameters?: Record<string, string[]>;
  frequencies?: string[];
  lifecycleStates?: string[];
  ruleStatuses?: string[];
  statuses: string[];
  severities: string[];
  origins: string[];
  actions: string[];
}

// Compiled-in copy of the backend catalog (app/catalog.py). It is what the
// filter and authoring dropdowns fall back to when /api/catalog/meta cannot be
// reached — an unreachable API should degrade to "no cases found", never to
// empty dropdowns the user cannot act on.
export const FALLBACK_CATALOG: CatalogMeta = {
  assurances: [
    { code: "RA", name: "Rating Assurance", group: "Commercial Assurance",
      modules: ["Rating", "Tariff", "Discount", "Usage Events", "Subscriber"] },
    { code: "PA", name: "Partner Assurance", group: "Commercial Assurance",
      modules: ["Partner", "Settlement", "Interconnect", "Roaming", "Invoice"] },
    { code: "MA", name: "Migration Assurance", group: "Customer Assurance",
      modules: ["Subscriber", "Account", "Balance", "Product Catalog", "Legacy Feed"] },
    { code: "UA", name: "Usage Assurance", group: "Revenue Assurance",
      modules: ["Usage Events", "Subscriber", "MSC", "CDR", "Mediation", "Rating", "Billing"] },
    { code: "BA", name: "Billing Assurance", group: "Revenue Assurance",
      modules: ["Invoice", "Bill Run", "Account", "Charge", "Adjustment", "Tax"] },
    { code: "CA", name: "Charging Assurance", group: "Revenue Assurance",
      modules: ["Balance", "Session", "OCS", "Voucher", "Subscriber", "Bundle"] },
    { code: "NA", name: "Network Assurance", group: "Operations Assurance",
      modules: ["Node", "MSC", "SGSN", "Probe", "Element Feed"] },
    { code: "CL", name: "Collection Assurance", group: "Financial Assurance",
      modules: ["Payment", "Receivable", "Dunning", "Account", "Bank Feed"] },
    { code: "ME", name: "Mediation Assurance", group: "Operations Assurance",
      modules: ["AIR", "SDP", "MSC", "CDR", "File Feed", "Mediation"] },
  ],
  groups: [
    "Commercial Assurance", "Customer Assurance", "Financial Assurance",
    "Operations Assurance", "Revenue Assurance",
  ],
  modules: [],  // derived from the assurances above where needed
  ruleCategories: [
    "Completeness", "Reconciliation", "Aggregation", "Threshold", "Sequence",
    "Duplicate", "Existence", "Statistical", "Temporal", "Comparison",
    "Calculation", "Referential Integrity", "Pattern Matching", "ML Prediction",
    "Graph Relationship", "Manual Investigation",
  ],
  frequencies: ["Real-time", "Hourly", "Cycle", "Daily", "Weekly", "Monthly"],
  lifecycleStates: ["Draft", "Active", "Paused", "Retired"],
  ruleStatuses: ["PASS", "FAIL", "WARNING", "NOT_RUN"],
  statuses: [...STATUSES],
  severities: [...SEVERITIES],
  origins: [...ORIGINS],
  actions: [...ACTIONS],
};

/** Every filter the list, summary and CSV export accept. */
export interface CaseQuery {
  q?: string;
  assurance?: string[];
  group?: string[];
  module?: string[];
  category?: string[];
  status?: string[];
  severity?: string[];
  origin?: string[];
  action?: string[];
  owner?: string[];
  ruleId?: string[];
  stream?: string[];
  assignment?: "all" | "assigned" | "unassigned" | "mine";
  me?: string;
  dateFrom?: string;
  dateTo?: string;
  dateField?: "createdAt" | "detectedAt" | "updatedAt";
  openOnly?: boolean;
  page?: number;
  pageSize?: number;
  sortBy?: string;
  sortDir?: "asc" | "desc";
}

// ---------------------------------------------------------------------------
// Transport
// ---------------------------------------------------------------------------

export function buildQuery(params: CaseQuery): URLSearchParams {
  const search = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === null || value === "") continue;
    // Array filters repeat the key — FastAPI reads them as a list.
    if (Array.isArray(value)) {
      for (const v of value) if (v !== "") search.append(key, String(v));
    } else {
      search.append(key, String(value));
    }
  }
  return search;
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  // FormData must set its own Content-Type: the browser appends the multipart
  // boundary, and overriding it here would make the body unparseable.
  const isForm = typeof FormData !== "undefined" && init?.body instanceof FormData;
  const res = await fetch(`${API_BASE}${path}`, {
    ...init,
    headers: {
      accept: "application/json",
      ...(init?.body && !isForm ? { "Content-Type": "application/json" } : {}),
      ...(init?.headers ?? {}),
    },
  });
  if (!res.ok) {
    // FastAPI puts the reason in `detail` — surface it rather than a bare code.
    let message = `${res.status} ${res.statusText}`;
    try {
      const body = await res.json();
      const detail = body?.detail;
      if (typeof detail === "string") message = detail;
      else if (Array.isArray(detail) && detail[0]?.msg) message = detail[0].msg;
    } catch {
      /* non-JSON error body */
    }
    throw new Error(message);
  }
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

// ---------------------------------------------------------------------------
// Reads
// ---------------------------------------------------------------------------

export const listCases = (query: CaseQuery = {}) =>
  request<CaseListResponse>(`/api/cases?${buildQuery(query)}`);

export const fetchSummary = (query: CaseQuery = {}) =>
  request<CaseSummary>(`/api/cases/summary?${buildQuery(query)}`);

export const fetchFacets = () => request<CaseFacets>("/api/cases/facets");

export const fetchCatalog = () => request<CatalogMeta>("/api/catalog/meta");

/** One control rule as authored in the rules module. */
export interface ControlRule {
  id: string;               // UA001
  name: string;
  intent: string;
  assuranceCode: string;
  assuranceName: string;
  assuranceGroup: string;
  entityScope: string;      // the module the rule is bound to
  primitiveCategory: string;
  severity: string;
  frequency: string;
  sourceFeed: string;
  targetFeed: string;
  tolerancePct: number | null;
  params: Record<string, unknown>;
  lifecycleState: string;
  lastStatus: string;
  lastRunAt: string | null;
  casesRaised: number;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
  openCases: number;
}

/** The provisioned controls — used to label the Rule filter and to trace a
 *  case back to the control that raised it. */
export const fetchRules = (params: {
  assurance?: string[]; category?: string[]; q?: string; pageSize?: number;
} = {}) =>
  request<{ items: ControlRule[]; total: number; page: number; pageSize: number; pageCount: number }>(
    `/api/rules?${buildQuery(params as CaseQuery)}`,
  );

/** One case with its mismatch rows, notes and audit trail. */
export const fetchCase = (id: string) =>
  request<AssuranceCase>(`/api/cases/${encodeURIComponent(id)}`);

export const fetchMismatches = (id: string, page = 1, pageSize = 100) =>
  request<{ items: CaseMismatch[]; total: number; page: number; pageSize: number; affectedCount: number }>(
    `/api/cases/${encodeURIComponent(id)}/mismatches?page=${page}&pageSize=${pageSize}`,
  );

/** URL for the server-side CSV export of the current filter set. */
export const exportCsvUrl = (query: CaseQuery = {}) =>
  `${API_BASE}/api/cases/export.csv?${buildQuery(query)}`;

// ---------------------------------------------------------------------------
// Writes
// ---------------------------------------------------------------------------

/** Payload for the Add Case dialog. `assurance` takes a code or a full name. */
export interface CreateCasePayload {
  title: string;
  description?: string;
  assurance: string;
  module?: string;
  subModule?: string;
  ruleId?: string | null;
  ruleName?: string | null;
  ruleCategory?: string;
  severity?: string;
  status?: string;
  action?: string;
  owner?: string;
  stream?: string;
  nodeId?: string;
  sourceFeed?: string;
  targetFeed?: string;
  linkedBatch?: string;
  linkedTxnId?: string;
  expectedValue?: string | null;
  actualValue?: string | null;
  variance?: string | null;
  threshold?: string | null;
  estimatedImpact?: number;
  affectedCount?: number;
  createdBy?: string;
  mismatches?: Partial<CaseMismatch>[];
}

export const createCase = (payload: CreateCasePayload) =>
  request<AssuranceCase>("/api/cases", { method: "POST", body: JSON.stringify(payload) });

export const updateCase = (
  id: string,
  payload: Partial<{
    title: string; description: string; status: string; severity: string;
    action: string; owner: string; module: string; subModule: string;
    tags: string[]; actor: string;
  }>,
) =>
  request<AssuranceCase>(`/api/cases/${encodeURIComponent(id)}`, {
    method: "PATCH",
    body: JSON.stringify({ actor: CURRENT_ANALYST, ...payload }),
  });

export const assignCase = (id: string, owner: string) =>
  request<AssuranceCase>(`/api/cases/${encodeURIComponent(id)}/assign`, {
    method: "POST",
    body: JSON.stringify({ owner, actor: CURRENT_ANALYST }),
  });

export const setCaseStatus = (id: string, status: string, action?: string, note?: string) =>
  request<AssuranceCase>(`/api/cases/${encodeURIComponent(id)}/status`, {
    method: "POST",
    body: JSON.stringify({ status, action, note, actor: CURRENT_ANALYST }),
  });

export const addComment = (id: string, body: string, author = CURRENT_ANALYST) =>
  request<CaseComment>(`/api/cases/${encodeURIComponent(id)}/comments`, {
    method: "POST",
    body: JSON.stringify({ author, body }),
  });

export const addInsight = (id: string, body: string) =>
  request<AssuranceCase>(`/api/cases/${encodeURIComponent(id)}/insights`, {
    method: "POST",
    body: JSON.stringify({ body }),
  });

export const deleteCase = (id: string) =>
  request<{ ok: boolean; message: string }>(`/api/cases/${encodeURIComponent(id)}`, {
    method: "DELETE",
  });

// ---------------------------------------------------------------------------
// Attachments
// ---------------------------------------------------------------------------

/** Only PDFs are accepted; the server re-checks both the type and the bytes. */
export const ACCEPTED_UPLOAD_TYPES = ["application/pdf"];
export const MAX_UPLOAD_BYTES = 25 * 1024 * 1024;   // mirrors MAX_UPLOAD_MB

/** Client-side pre-check, so an obvious mistake doesn't need a round trip. */
export function validateUploadFile(file: File): string | null {
  const isPdf = file.type === "application/pdf" || file.name.toLowerCase().endsWith(".pdf");
  if (!isPdf) return "Only PDF files can be attached.";
  if (file.size === 0) return "That file is empty.";
  if (file.size > MAX_UPLOAD_BYTES) return `File is larger than the ${MAX_UPLOAD_BYTES / 1024 / 1024} MB limit.`;
  return null;
}

export const uploadAttachments = async (
  caseId: string, files: File[], uploadedBy = CURRENT_ANALYST,
): Promise<CaseAttachment[]> => {
  const form = new FormData();
  for (const file of files) form.append("files", file);
  form.append("uploadedBy", uploadedBy);
  // No Content-Type header: the browser has to set the multipart boundary.
  return request<CaseAttachment[]>(`/api/cases/${encodeURIComponent(caseId)}/attachments`, {
    method: "POST",
    body: form,
  });
};

export const listAttachments = (caseId: string) =>
  request<CaseAttachment[]>(`/api/cases/${encodeURIComponent(caseId)}/attachments`);

export const deleteAttachment = (caseId: string, attachmentId: string) =>
  request<{ ok: boolean; message: string }>(
    `/api/cases/${encodeURIComponent(caseId)}/attachments/${encodeURIComponent(attachmentId)}?actor=${encodeURIComponent(CURRENT_ANALYST)}`,
    { method: "DELETE" },
  );

/** Absolute URL for an attachment — the API returns a relative path. */
export const attachmentUrl = (attachment: CaseAttachment) => `${API_BASE}${attachment.downloadUrl}`;

export const fmtBytes = (bytes: number) => {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(0)} KB`;
  return `${(bytes / 1024 / 1024).toFixed(1)} MB`;
};

// ---------------------------------------------------------------------------
// Display helpers
// ---------------------------------------------------------------------------

export const isUnassigned = (c: AssuranceCase) => !c.owner.trim();
export const ownerLabel = (c: AssuranceCase) => (isUnassigned(c) ? "Unassigned" : c.owner);

/** "UA001 · Missing CDR", or a dash for an analyst-raised case. */
export const ruleLabel = (c: AssuranceCase) =>
  c.ruleId ? `${c.ruleId}${c.ruleName ? ` · ${c.ruleName}` : ""}` : "—";

/** The headline mismatch, e.g. "1,284,322 → 1,281,904 (-2,418)". */
export function mismatchLabel(c: AssuranceCase): string {
  if (c.expectedValue == null && c.actualValue == null) return "—";
  const arrow = `${c.expectedValue ?? "—"} → ${c.actualValue ?? "—"}`;
  return c.variance ? `${arrow} (${c.variance})` : arrow;
}

export const fmtMoney = (n: number) => `$${new Intl.NumberFormat("en-US").format(Math.round(n))}`;

export const fmtDate = (iso: string) =>
  new Date(iso).toLocaleString(undefined, {
    day: "2-digit", month: "short", year: "numeric", hour: "2-digit", minute: "2-digit",
  });

export const fmtDay = (iso: string) =>
  new Date(iso).toLocaleDateString(undefined, { day: "2-digit", month: "short", year: "numeric" });

/** "2h ago" — derived at render from a real timestamp, never stored. */
export function relative(isoStr: string): string {
  const ms = Date.now() - new Date(isoStr).getTime();
  const mins = Math.floor(ms / 60000);
  if (mins < 1) return "just now";
  if (mins < 60) return `${mins}m ago`;
  const hrs = Math.floor(mins / 60);
  if (hrs < 24) return `${hrs}h ago`;
  return `${Math.floor(hrs / 24)}d ago`;
}
